### DISCUSSION TITLE
Unable to inspect star icon IDs in Appium Inspector

### DISCUSSION THREAD

**Discussion by RoeyP** [appium/appium#20766] [Q&A]
I'm experiencing an issue with Appium Inspector where I'm unable to inspect the IDs of star rating elements in my iOS mobile app, which is written in SwiftUI. The IDs in question were added by the developer in the format `rc_rating_star_view_<#>`, e.g. `rc_rating_star_view_5` for a 5-star rating.

While the star rating elements are visible in the display area of the Appium Inspector, I'm unable to find their IDs in the XML hierarchy. I've checked the entire XML structure, but the IDs are nowhere to be found. Other IDs in the app are accessible, but these specific IDs are not.

I've verified that the IDs are present in the app's code, and I'm able to access other elements using Appium. However, I'm stuck on this specific issue.

**Additional information:**

* Appium version: 7.6.0
* Mobile platform and version: iOS17.0 / iPhone 15
* Appium Inspector version: 2024.9.1 

**Question:** Has anyone else encountered this issue, or can anyone provide guidance on how to inspect these star rating IDs in Appium Inspector?

Screenshot:
<img width="1042" alt="Screenshot 2024-11-19 at 16 15 47" src="https://github.com/user-attachments/assets/42da1a11-d296-4032-91ff-2a14399e1c5e">

**Comment by portyu9** [comment]
Hi @RoeyP 

If the identifier is not present in Appium's XML/page source, Appium cannot locate it. This needs to be fixed in the SwiftUI accessibility hierarchy.

Apply the identifier to the **actual actionable accessibility element**, preferably the `Button`, rather than only to a decorative star `Image`:

```swift
Button {
    rating = 5
} label: {
    Image(systemName: "star.fill")
}
.accessibilityIdentifier("rc_rating_star_view_5")
.accessibilityLabel("5 stars")
```

If a parent SwiftUI view is merging the stars into one accessibility element, preserve the individual children:

```swift
.accessibilityElement(children: .contain)
```

After rebuilding the app, refresh Appium Inspector and confirm that each `rc_rating_star_view_<n>` appears in the XML.

The fix is on the SwiftUI accessibility structure; changing the Appium locator will not help while the individual stars are absent from XCTest's hierarchy.

Please mark this answer as accepted if it helped, thank you.
